implementation/src/main/java/io/smallrye/config/ProfileConfigSourceInterceptor.java

```diff
@@ -0,0 +1,39 @@
+package io.smallrye.config;
+
+import javax.annotation.Priority;
+
+import org.eclipse.microprofile.config.Config;
+
+@Priority(600)
+public class ProfileConfigSourceInterceptor implements ConfigSourceInterceptor {
+    public static final String SMALLRYE_PROFILE = "smallrye.config.profile";
+
+    private String profile;
+
+    public ProfileConfigSourceInterceptor() {
+
+    }
+
+    public ProfileConfigSourceInterceptor(final String profile) {
+        this.profile = profile;
+    }
+
+    @Override
+    public void initialize(final Config config) {
+        if (profile == null) {
+            profile = config.getOptionalValue(SMALLRYE_PROFILE, String.class).orElse(null);
+        }
+    }
+
+    @Override
+    public ConfigValue getValue(final ConfigSourceInterceptorContext context, final String name) {
+        if (profile != null) {
+            final ConfigValue profileValue = context.proceed("%" + profile + "." + name);
+            if (profileValue != null) {
+                return profileValue;
```

Comment:
This is slightly incorrect unfortunately.

The following cases should be tested.  Given ConfigSource A(ordinal 100) and B(ordinal 200):

* _A_ contains `%profile.foo` and _B_ contains `foo`: _A_ wins
* _A_ contains `foo` and _B_ contains `%profile.foo`: _A_ still wins
* _A_ contains both `%profile.foo` and `foo`: _A_ `%profile.foo` wins

Otherwise we have a situation where a config source of higher ordinal can override one of lower ordinal.

So this means that I was incorrect in my original assumption:  We have to get `ConfigValue` instances for both property names, and then if both are returned, we have to compare the ordinals and return the value for the source with the lower ordinal.  Good thing we put ordinal on `ConfigValue`, right? :wink:

Reply:
Hehe, yes. 

I guess I just kept that code we discussed some time ago in my mind and didn't think too much about it.

Let me test and fix that.

Reply:
BTW,

> we have to compare the ordinals and return the value for the source with the lower ordinal.

Shouldn't this be:

return the value for the source with the **higher** ordinal?

Reply:
Here's my trick to remember ordinal vs priority:

* Higher priority wins (think "this is top priority!")
* Lower ordinal wins (think "1st is more important than 2nd")

Reply:
If it were my choice though, I would have used priority for config source as well...

Reply:
One clarification:

In:
> A contains `foo` and B contains `%profile.foo`: A still wins

Why don't we want to have B override A if B has a higher priority and it is the active profile?

Reply:
> * Lower ordinal wins (think "1st is more important than 2nd")

Sure, but in the `getOrdinal` java docs we have:

```java
* If a property is specified in multiple config sources, the value in the config source with the highest ordinal
* takes precedence.
```

If this is the case, shouldn't it be `lowest`?

Reply:
Because then if users want to override a value (for example using a system property), they must now know whether the value came from a profile (and which profile might be active) before they can do so.  But what users generally *want* is that if they set a system property or env var, that takes precedence no matter what.  This is consistent with the behavior of Quarkus today.

Reply:
> > * Lower ordinal wins (think "1st is more important than 2nd")
> 
> Sure, but in the `getOrdinal` java docs we have:
> 
> ```java
> * If a property is specified in multiple config sources, the value in the config source with the highest ordinal
> * takes precedence.
> ```
> 
> If this is the case, shouldn't it be `lowest`?

Yes, historically this was mixed up in a *number* of places, and I see we've mixed it up again. :-)

Reply:
Well, then it is a huge mix. Because even the default ConfigSources use the notion of priority:

```java
 * <li>{@linkplain System#getProperties() System properties}, with an {@linkplain #getOrdinal() ordinal value} of {@code 400}</li>
 * <li>{@linkplain System#getenv() Environment properties}, with an ordinal value of {@code 300}</li>
 * <li>The {@code /META-INF/microprofile-config.properties} {@linkplain ClassLoader#getResource(String) resource}, 
 * with an ordinal value of {@code 100}</li>
```

And we don't want `microprofile-config.properties` to be the most important :)

Reply:
> Because then if users want to override a value (for example using a system property), they must now know whether the value came from a profile (and which profile might be active) before they can do so. But what users generally _want_ is that if they set a system property or env var, that takes precedence no matter what. This is consistent with the behavior of Quarkus today.

Effectively, it might just be easier to define that profiles are applied per ConfigSource?

Reply:
> Well, then it is a huge mix. Because even the default ConfigSources use the notion of priority:

Yeah maybe I have it backwards!  Mixing things up is sadly common for me.  Maybe we should change all incidence of "ordinal" to "priority" and end the confusion for good (even in the specification)!

> > Because then if users want to override a value (for example using a system property), they must now know whether the value came from a profile (and which profile might be active) before they can do so. But what users generally _want_ is that if they set a system property or env var, that takes precedence no matter what. This is consistent with the behavior of Quarkus today.
> 
> Effectively, it might just be easier to define that profiles are applied per ConfigSource?

It would be easier to implement, but not useful to users to do it this way.  The config source priority has to be considered before the profile.

Reply:
Just trying to figure out the best way to then explain to the users. In Quarkus documentation this does not seem clear:
https://quarkus.io/guides/config#configuration-profiles

So we want to say that higher priority (or lower ordinal :)) sources have precedence over a profile configuration.

Reply:
Yes, I agree with that.  Sorry if I didn't understand you.

Reply:
No worries. I most likely I didn't explain myself properly :)